ostree/ostree_dest.go

```diff
@@ -0,0 +1,245 @@
+func (d *ostreeImageDestination) Close() {
+}
+
+func (d *ostreeImageDestination) SupportedManifestMIMETypes() []string {
+	return []string{
+		manifest.DockerV2Schema2MediaType,
+	}
+}
+
+// SupportsSignatures returns an error (to be displayed to the user) if the destination certainly can't store signatures.
+// Note: It is still possible for PutSignatures to fail if SupportsSignatures returns nil.
+func (d *ostreeImageDestination) SupportsSignatures() error {
+	return nil
+}
+
+// ShouldCompressLayers returns true iff it is desirable to compress layer blobs written to this destination.
+func (d *ostreeImageDestination) ShouldCompressLayers() bool {
+	return false
+}
+
+// AcceptsForeignLayerURLs returns false iff foreign layers in manifest should be actually
+// uploaded to the image destination, true otherwise.
+func (d *ostreeImageDestination) AcceptsForeignLayerURLs() bool {
+	return false
+}
+
+func (d *ostreeImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobInfo) (types.BlobInfo, error) {
+	hash := strings.Replace(inputInfo.Digest.Hex(), "sha256:", "", 1)
+
+	blobPath := filepath.Join(encodeOStreeRef(d.ref.image), hash, "content")
+	err := ensureParentDirectoryExists(blobPath)
+	if err != nil {
+		return types.BlobInfo{}, err
+	}
+
+	blobFile, err := os.Create(blobPath)
+	if err != nil {
+		return types.BlobInfo{}, err
+	}
+	defer func() {
+		blobFile.Close()
+	}()
+
+	digester := digest.Canonical.Digester()
+	tee := io.TeeReader(stream, digester.Hash())
+
+	size, err := io.Copy(blobFile, tee)
+	if err != nil {
+		return types.BlobInfo{}, err
+	}
+	computedDigest := digester.Digest()
+	if inputInfo.Size != -1 && size != inputInfo.Size {
+		return types.BlobInfo{}, errors.Errorf("Size mismatch when copying %s, expected %d, got %d", computedDigest, inputInfo.Size, size)
+	}
+	if err := blobFile.Sync(); err != nil {
+		return types.BlobInfo{}, err
+	}
+
+	d.blobs[hash] = &blobToImport{Size : size, Digest : inputInfo.Digest, BlobPath : blobPath, Hash : hash}
+	return types.BlobInfo{Digest: computedDigest, Size: size}, nil
+}
+
+func (d *ostreeImageDestination) importBlob(isTar bool, blob *blobToImport) (error) {
+	ostreeBranch := fmt.Sprintf("ociimage/%s", blob.Hash)
+
+	if ! isTar {
+		return exec.Command("ostree", "commit",
+			fmt.Sprintf("--add-metadata-string=docker.size=%d", blob.Size),
+			fmt.Sprintf("--branch=%s", ostreeBranch), filepath.Dir(blob.BlobPath)).Run()
+	}
+
+	destinationPath := filepath.Join(encodeOStreeRef(d.ref.image), blob.Hash, "root")
+	if err := ensureDirectoryExists(destinationPath); err != nil {
+		return err
+	}
+	defer func() {
+		os.Remove(blob.BlobPath)
+		os.RemoveAll(destinationPath)
+	}()
+
+	err := archive.UntarPath(blob.BlobPath, destinationPath)
+	if err != nil {
+		return err
+	}
+	return exec.Command("ostree", "commit",
+		fmt.Sprintf("--add-metadata-string=docker.size=%d", blob.Size),
+		fmt.Sprintf("--branch=%s", ostreeBranch),
+		fmt.Sprintf("--tree=dir=%s", destinationPath)).Run()
```

Comment:
(Just to be sure, is the untarred filesystem at `destinationPath` supposed to exist after this command returns, or is it supposed to be temporary?)